cas-server-core/src/main/java/org/jasig/cas/authentication/support/DefaultCasAttributeEncoder.java

```diff
@@ -0,0 +1,226 @@
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL);
+    }
+
+    /**
+     * Encode and encrypt pgt using the public key
+     * supplied by the service. The result is base64 encoded
+     * and put into the attributes collection again, overwriting
+     * the previous value.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     */
+    protected void encodeAndEncryptProxyGrantingTicket(final Map<String, Object> attributes,
+                                                       final Map<String, String> cachedAttributesToEncode) {
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);
+    }
+
+    /**
+     * Encrypt, encode and put the attribute into attributes map.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param cachedAttributeName the cached attribute name
+     */
+    protected void encryptAndEncodeAndPutIntoAttributesMap(final Map<String, Object> attributes,
+                                                           final Map<String, String> cachedAttributesToEncode,
+                                                           final String cachedAttributeName) {
+        final String cachedAttribute = cachedAttributesToEncode.remove(cachedAttributeName);
+        if (StringUtils.isNotBlank(cachedAttribute)) {
+            logger.debug("Retrieved [{}] as a cached model attribute...", cachedAttributeName);
+            final String encodedValue = CompressionUtils.encryptAndEncodeBase64(cachedAttribute, this.cipher);
+            attributes.put(cachedAttributeName, encodedValue);
+            logger.debug("Encrypted and encoded [{}] as an attribute to [{}].", cachedAttributeName, encodedValue);
+        } else {
+            logger.debug("[{}] is not available as a cached model attribute to encrypt...", cachedAttributeName);
+        }
+    }
+
+    /**
+     * Initialize the cipher with the public key
+     * and then start to encrypt select attributes.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param service the service
+     */
+    protected void encodeAttributesInternal(final Map<String, Object> attributes,
+                                            final Map<String, String> cachedAttributesToEncode,
+                                            final RegisteredService service) {
+        try {
+            if (!initializeCipherBasedOnServicePublicKey(service)) {
+                return;
+            }
+            encodeAndEncryptCredentialPassword(attributes, cachedAttributesToEncode);
```

Comment:
I'm not sure to follow you here.

I guess the ```encodeAttributes``` method  should be ```final``` because otherwise, I can override it and get directly the password and PGT, can't I?

Attributes are provided in plain text, so we must absolutely be sure that the password and PGT are encrypted and with the current design, it means we must be sure that they have been saved in the ```cachedAttributesToEncode```, which is performed by the ```initialize``` method which thus, should be marked ```final``` as well.

Am I wrong?